app/service/learning_svc.py

```diff
@@ -35,43 +37,49 @@ async def build_model(self):
                         self.model.add(variables)
         self.model = set(self.model)
 
-    async def learn(self, facts, link, blob):
+    async def learn(self, facts, link, blob, operation=None):
         decoded_blob = b64decode(blob).decode('utf-8')
 
         found_facts = []
         for parser in self.parsers:
             try:
                 for fact in parser.parse(decoded_blob):
-                    await self._save_fact(link, facts, fact)
                     found_facts.append(fact)
             except Exception as e:
                 self.log.error(e)
-        await self._update_scores(link, facts, increment=len(found_facts))
-        await self._build_relationships(link, found_facts)
+        await _update_scores(operation=None, increment=len(found_facts), used=facts, facts=link.facts)
+        await self._build_relationships(link, found_facts, operation)
 
     """ PRIVATE """
 
     @staticmethod
-    async def _update_scores(link, facts, increment):
-        for uf in link.facts:
-            for found_fact in facts:
-                if found_fact.unique == uf.unique:
-                    found_fact.score += increment
-                    break
-
-    @staticmethod
-    async def _save_fact(link, facts, fact):
-        if all(fact.trait) and not any(f.trait == fact.trait and f.value == fact.value for f in facts):
+    async def _save_fact(link, facts, fact, operation=None):
+        if all(fact.trait) and not any(f.trait == fact.trait and f.value == fact.value and f.source == link.id
+                                       for f in facts):
             fact.collected_by = link.paw
             fact.technique_id = link.ability.technique_id
+            fact.links = [link]
+            fact.relationships = []
+            fact.source_type = OriginType.LEARNED.name
+            fact.source = operation.id if operation else link.id
+            knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+            await knowledge_svc_handle.add_fact(fact)
             link.facts.append(fact)
 
-    async def _build_relationships(self, link, facts):
+    async def _build_relationships(self, link, facts, operation=None):
+        facts_covered = []
         for relationship in self.model:
             matches = []
             for fact in facts:
                 if fact.trait in relationship:
                     matches.append(fact)
+                    facts_covered.append(fact)
+                else:
+                    await self._save_fact(link, facts, fact)
             for pair in itertools.combinations(matches, r=2):
                 if pair[0].trait != pair[1].trait:
-                    link.relationships.append(Relationship(source=pair[0], edge='has', target=pair[1]))
+                    await link._create_relationships([Relationship(source=pair[0], edge='has', target=pair[1])],
```

Comment:
It looks like we're saving facts by calling the method on the learning service and passing in the link, but saving relationships by calling a method on the link class. The link has a `_save_fact()` method; should it be used instead? What's the difference?

Reply:
I think originally I was trying to do something fancy to make it obvious whether or not a learning parser was utilized to create a fact, but that never really panned out. Combining them makes sense at this point.